### ISSUE TITLE
Path not being fixed

### ISSUE THREAD

**Issue by davej** [sindresorhus/fix-path#4]
```
> echo $PATH
/Library/Frameworks/Python.framework/Versions/3.4/bin:/Applications/Postgres.app/Contents/Versions/9.3/bin:/usr/local/heroku/bin:/usr/local/sbin:/Users/dave/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/deployd/bin:/usr/local/go/bin:/usr/local/MacGPG2/bin
```

Using Electron 0.30.2:

``` javascript
require('fix-path')()
var dialog = require('dialog')

dialog.showErrorBox("path", process.env.PATH)
```

![screen shot 2015-07-31 at 18 50 39](https://cloud.githubusercontent.com/assets/26971/9013713/950f5fc8-37b5-11e5-8f82-01ff1651fc7c.png)

It works fine on my Macbook Pro but fails to list all the paths on my Macbook Air (as shown above). They are both running Yosemite. I have also reproduced the same issue on a friends Macbook.

Any idea what could be going on here?

**Comment by silverwind** [comment]
Which shell is that and where are those path extensions set? Also, is `shell-path` at 1.0.2?

**Comment by davej** [comment]
Just the normal OS X bash shell, I run it with terminal. Most of the paths are set in `~/.profile` but there are a couple of others which must be coming from somewhere else (is there a way to find out where they are being set from?).

Yes, `shell-path` is 1.0.2.

**Comment by silverwind** [comment]
I assume it's bash, but could you post `echo $SHELL`, just to be sure?

Most paths probably come from `/private/etc/profile`.

**Comment by davej** [comment]
```
> echo $SHELL
/bin/bash
```

**Comment by silverwind** [comment]
is it `/bin/bash` in your electron app too?

``` js
dialog.showErrorBox("path", process.env.SHELL)
```

**Comment by davej** [comment]
Yup.

![screen shot 2015-08-01 at 14 34 51](https://cloud.githubusercontent.com/assets/26971/9022015/95534094-385a-11e5-9a5c-30e74c1efacf.png)

**Comment by silverwind** [comment]
Hmm, hard one. What about

``` js
dialog.showErrorBox("path", require("child_process").execSync('/bin/bash -i -c "echo $PATH"').toString())
```

**Comment by davej** [comment]
Same as before: `/usr/bin:/bin:/usr/sbin:/sbin`.

**Comment by silverwind** [comment]
Hmm, I honestly have no idea. Does `/bin/bash -i -c "echo $PATH"` in terminal give the correct path?

**Comment by davej** [comment]
Yes. 

```
> /bin/bash -i -c "echo $PATH"
/Library/Frameworks/Python.framework/Versions/3.4/bin:/Applications/Postgres.app/Contents/Versions/9.3/bin:/usr/local/heroku/bin:/usr/local/sbin:/Users/dave/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/deployd/bin:/usr/local/go/bin:/usr/local/MacGPG2/bin
```

Hmmm, it's a strange one alright. I have recreated this behaviour on two macbook airs (my own and my friends). They were originally OS X 10.7 and have been upgraded through the years to 10.10. Perhaps some obscure config got mangled in one of the upgrades or something?

**Comment by silverwind** [comment]
Does it work in node?

``` console
node -p 'require("child_process").execSync("/bin/bash -i -c \"echo $PATH\"").toString()'
```

**Comment by davej** [comment]
Yup, works from node.

```
> node -p 'require("child_process").execSync("/bin/bash -c \"echo $PATH\"").toString()'
/Library/Frameworks/Python.framework/Versions/3.4/bin:/Applications/Postgres.app/Contents/Versions/9.3/bin:/usr/local/heroku/bin:/usr/local/sbin:/Users/dave/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/deployd/bin:/usr/local/go/bin:/usr/local/MacGPG2/bin
```

**Comment by davej** [comment]
Updated to latest iojs just in case (system node was 0.11). Same thing:

```
>  node -p 'require("child_process").execSync("/bin/bash -c \"echo $PATH\"").toString()'
/Users/dave/.nvm/versions/io.js/v2.5.0/bin:/Library/Frameworks/Python.framework/Versions/3.4/bin:/Applications/Postgres.app/Contents/Versions/9.3/bin:/usr/local/heroku/bin:/usr/local/sbin:/Users/dave/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/deployd/bin:/usr/local/go/bin:/usr/local/MacGPG2/bin
```

**Comment by silverwind** [comment]
Hmm is it even executing the same bash?

``` js
dialog.showErrorBox("path", require("child_process").execFileSync("/bin/bash", ["--version"]).toString())
```

```
node -p 'require("child_process").execFileSync("/bin/bash", ["--version"]).toString()'
```

**Comment by davej** [comment]
This also worked by the way (running from bash): `> ~/Downloads/Test.app/Contents/MacOS/Electron`

![screen shot 2015-08-01 at 15 05 37](https://cloud.githubusercontent.com/assets/26971/9022102/c91d3a66-385e-11e5-9c1d-2ccd4388ecd7.png)

**Comment by davej** [comment]
```
>  node -p 'require("child_process").execFileSync("/bin/bash", ["--version"]).toString()'
GNU bash, version 3.2.53(1)-release (x86_64-apple-darwin14)
Copyright (C) 2007 Free Software Foundation, Inc.
```

![screen shot 2015-08-01 at 15 09 31](https://cloud.githubusercontent.com/assets/26971/9022108/57973c24-385f-11e5-946f-1bb1e000a03d.png)

**Comment by silverwind** [comment]
I'm out of ideas, electron seems to do something differently with bash, but I have no idea what. Maybe you could try downgrading electron if there was a recent change or something.

**Comment by davej** [comment]
Ok, thanks for your help. For the moment, I'll just explicitly add potential paths to `process.env.PATH` and hope that covers it. Just to reiterate, the paths are working fine with Electron on my Macbook Pro which makes me think it's something to do with the OS X config on the Macbook Air.

I'll try an older version of Electron too. Thanks again.

**Comment by silverwind** [comment]
Should be fixed as of `shell-path` 1.0.4. @davej could you reinstall your modules and retest?

**Comment by davej** [comment]
I don't have the macbook air anymore so I won't be able to retest. I guess we can probably close this issue though if you think it should be fixed?

**Comment by silverwind** [comment]
I was never able to reproduce on my machine, but I'm pretty sure it's fixed, thanks.

**Comment by davej** [comment]
I received a report of this issue again. They were using fix-path 1.1.0 and shell-path 1.0.4. I stopped receiving these errors because I had manually added in common $PATHs (e.g. `/usr/local/bin`). I took the hack out when I updated to the latest shell-path and the errors are back now.

I may be able to do some debugging of this on their machine on Thursday, so if you have any suggestions then let me know.

**Comment by silverwind** [comment]
I'd be interested in the output of

``` js
require('child_process').execSync('sudo -Hiu' + process.env.USER + ' echo $PATH', {encoding: 'utf8'})
```

Reading through https://github.com/atom/electron/issues/550 makes me believe the issue only manifests when the application is not started through a terminal (e.g. through `open` when its started through the GUI).

**Comment by davej** [comment]
After some testing on my local machine, I believe that the issue may be caused by usage of the `child_process` `execSync` method in the Electron renderer.

Basically, the first method (`pathFromShellSync`) runs ok but causes an error (`write EBADF`), the subsequent method (`pathFromSudoSync`) fails because the socket is closed (due to the first error). This only effects the renderer, not the main process.

See here: https://github.com/atom/electron/issues/3482

**Comment by silverwind** [comment]
Interesting. I don't suppose catching that error makes any difference?

Does `execFile` behave differently perhaps?

``` js
require('child_process').execFileSync('sudo', ['-Hiu', process.env.USER, 'echo "$PATH"'], {encoding: 'utf8'})
```

**Comment by davej** [comment]
No luck :(
`execFileSync` also fails. Catching the error is the same, the socket still closes.

**Comment by silverwind** [comment]
Do you have any idea which socket this is referring to? I'm not too familar with Electron :)

**Comment by davej** [comment]
Presumably the socket for streaming data to the child_process. I don't really understand it either though.

Here's the stacktrace: [net.js:640](https://github.com/nodejs/node/blob/master/lib/net.js#L640)

```
Uncaught Error: This socket is closed.(…)
Socket._writeGeneric @ net.js:640
Socket._write @ net.js:694
doWrite @ _stream_writable.js:292
writeOrBuffer @ _stream_writable.js:278
Writable.write @ _stream_writable.js:207
Socket.write @ net.js:618
execFileSync @ child_process.js:471(anonymous function) @ VM395:2
InjectedScript._evaluateOn @ VM184:904
InjectedScript._evaluateAndWrap @ VM184:837
InjectedScript.evaluate @ VM184:693
```

**Comment by silverwind** [comment]
Well, I guess you have to hardcode paths for now, this doesn't look like a simple fix. I'll try to dig deeper thought.

**Comment by davej** [comment]
Yeah, it's an upstream bug, I don't think there's much that can be done except wait for a fix. I might just use the `shell-path` async method instead.

**Comment by davej** [comment]
Perhaps a fix-path.async method might be useful though?

**Comment by silverwind** [comment]
Yeah, that'd be a worthwile addition I think, as `shell-path` already provides a async method. Care to file a PR?

**Comment by davej** [comment]
Sure.

**Comment by sindresorhus** [comment]
Closing as this was an old Electron bug from 2015 that has long since been fixed. The package has also gone through multiple major versions since then.
